### PR TITLE
feat[lang]: use keyword arguments for event instantiation

### LINKED ISSUE #4249: VIP: Convert Event construction to use kwargs
## Simple Summary
Make the syntax for creating event instances more clear, readable, and consistent with the syntax for struct creation by leveraging keyword arguments.

## Motivation
Structs have been updated to be constructed by kwargs ( #2381 ). This new syntax makes struct construction more convenient than the previous implementation.

Event creation could benefit from this same syntax change, not only improving the existing event syntax, but also making it consistent with struct creation, which lowers the mental overhead programmers need to work with when writing Vyper.

Right now, users have to remember the two different, but similar syntaxes for creating events and for creating structs. The syntax for creating structs is more readable, clear, and explicit. So we should update events to use the same syntax as structs.

## Specification
There are no semantic changes involved in this VIP.

### Curent Syntax
The current syntax around events uses positional arguments as follows
```vyper
# event definition
event EntryCreated:
    id: indexed(uint256)
    round_id: indexed(uint256)
    owner: indexed(address)
    amount: uint256

# event construction
log EntryCreated(foo, bar, msg.sender, 1000)
```
### Proposed Syntax
We propose changing this to the following syntax which uses keyword arguments
```vyper
# event definition (unchanged)
event EntryCreated:
    id: indexed(uint256)
    round_id: indexed(uint256)
    owner: indexed(address)
    amount: uint256

# event construction
log EntryCreated(id=foo, round_id=bar, owner=msg.sender, amount=1000)
```

### Comparison with Struct Syntax
The proposed kwarg is more consistent with the current syntax around structs, and the benefits of the change are very apparent when seen side by side

Given the following struct and event definitions:

```vyper
# Struct Definition
struct Entry:
    id: uint256
    round_id: uint256
    owner: address
    amount: uint256

# Event Definition (unchanged)
event EntryCreated:
    id: indexed(uint256)
    round_id: indexed(uint256)
    owner: indexed(address)
    amount: uint256
```

We construct these as follows:

**Existing Syntax**
```vyper
# Creating a Struct
entry: Entry = Entry(id=foo, round_id=bar, owner=msg.sender, amount=10000)
# Logging an Event
log EntryCreated(foo, bar, msg.sender, 10000)
```

**Proposed Syntax**
```vyper
# Creating a Struct
entry: Entry = Entry(id=foo, round_id=bar, owner=msg.sender, amount=10000)
# Logging an Event
log EntryCreated(id=foo, round_id=bar, owner=msg.sender, amount=10000)
```

## Backwards Compatibility
Similarly to the Struct Kwarg VIP, we can support both syntax options via an AST transformation and warn the user about the change. After a few releases, the previous syntax can be dropped, as it is better to have *one* syntax for this in the long run.

## Dependencies
N/A

## References
#2381 
#3777 

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

### PR DESCRIPTION
**Pull request by z80dev**
### What I did
Closes #4249 

Implement support for keyword arguments when calling event constructors

### How I did it
Mainly updated code to check `node.keywords` instead of `node.args` where dealing with events and the arguments passed to the constructor.

During semantic analysis, if we detect an event instantiated with positional args, we log a warning. Semantic analysis is the first place we can deal with this problem, as during parsing we don't have access to info such as what the names of a specific events' arguments are.

This implementation is fully backwards compatible with the positional arguments syntax.

### How to verify it

All unit tests have been updated to use keyword arguments for events, so passing tests give us a lot of confidence the change is correct.

In addition, a new unit test was added verifying that positional arguments still compile and work as expected.

### Commit message

```
This commit changes event instantiation from taking positional
arguments to taking keyword arguments.

For backwards compatibility (and to ease users into v0.4.1), this
implementation is fully backwards compatible with the existing
positional arguments syntax, but a warning will be emitted if the old
syntax is encountered.

All unit tests have been updated to use keyword arguments for events,
so passing tests give us a lot of confidence the change is correct.

In addition, a new unit test was added verifying that positional
arguments still compile and work as expected.

Misc/refactors:
- Factor out `validate_kwargs()` from `StructT._ctor_call_return()`
- Rewrite the kwarg validation code for clarity
- Add `__sub__` and `__isub__` dunders to `OrderedSet`
```

### Description for the changelog

Use kwargs for event instantiation

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/semantics/types/user.py [hunk 1]
```diff
@@ -281,9 +282,59 @@ def from_EventDef(cls, base_node: vy_ast.EventDef) -> "EventT":
...
+            msg += "in a future release. Use kwargs instead eg. "
+            msg += "Foo(a=1, b=2)"
+
+            vyper_warn(msg, node)
+
+            # convert positional args to keywords
+            kw_list = []
+            for kw, val in zip(self.arguments.keys(), node.args):
+                kw_node = vy_ast.keyword(arg=kw, value=val)
+                kw_node.value.set_parent(kw_node)  # type: ignore
```
Comment (tserg) [inline vyper/semantics/types/user.py:307]: I think this line is unnecessary.

**Review by tserg — COMMENTED** [review]
Left a comment, otherwise LGTM!

Handing it over to @charles-cooper

File: vyper/semantics/types/user.py (hunk 1 above)
Comment (z80dev) [reply vyper/semantics/types/user.py]: you're right! only the other `set_parent` call is needed (or else source maps break), I removed this extraneous line

File: vyper/semantics/types/user.py [hunk 2]
```diff
@@ -281,9 +282,58 @@ def from_EventDef(cls, base_node: vy_ast.EventDef) -> "EventT":
...
+        keys = list(self.arguments.keys())
+        for i, kwarg in enumerate(node.keywords):
+            # x=5 => kwarg(arg="x", value=Int(5))
+            argname = kwarg.arg
+            if argname not in members:
+                hint = get_levenshtein_error_suggestions(argname, members, 1.0)
+                raise UnknownAttribute("Unknown or duplicate event argument.", kwarg, hint=hint)
+            expected = keys[i]
+            if argname != expected:
+                raise InvalidAttribute(
```
Comment (charles-cooper) [inline vyper/semantics/types/user.py:324]: not covered by tests (https://github.com/vyperlang/vyper/pull/4257/checks?check_run_id=30669010870)

Comment (z80dev) [reply vyper/semantics/types/user.py:323]: added test covering these cases

File: vyper/codegen/stmt.py [hunk 3]
```diff
@@ -93,7 +93,13 @@ def parse_If(self):
...
         event = self.stmt._metadata["type"]
 
-        args = [Expr(arg, self.context).ir_node for arg in self.stmt.value.args]
+        args = []
+        if len(self.stmt.value.keywords) > 0:
+            # keyword arguments
+            args = [Expr(arg.value, self.context).ir_node for arg in self.stmt.value.keywords]
+        else:
+            # positional arguments
+            args = [Expr(arg, self.context).ir_node for arg in self.stmt.value.args]
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:102]: factor the code with effects (`Expr()` constructor) to the straight-line code section
```suggestion
        if len(self.stmt.value.keywords) > 0:
            # keyword arguments
            to_compile = [arg.value for arg in self.stmt.value.keywords]
        else:
            # positional arguments
            to_compile = self.stmt.value.args
        args = [Expr(arg, self.context).ir_node for arg in to_compile]
```

File: vyper/semantics/types/user.py [hunk 4]
```diff
@@ -281,10 +282,58 @@ def from_EventDef(cls, base_node: vy_ast.EventDef) -> "EventT":
...
+                    "Event keys are required to be in order, but got "
+                    f"`{argname}` instead of `{expected}`. (Reminder: the "
+                    f"keys in this event are {list(self.arguments)})",
+                    kwarg,
+                )
+            expected_type = members.pop(argname)
+            validate_expected_type(kwarg.value, expected_type)
```
Comment (charles-cooper) [inline vyper/semantics/types/user.py:330]: can we clean this code up? refactor it to be shared somehow with the struct code? if it ends up being more messy, we can declare it out of scope for this PR.

Comment (z80dev) [reply vyper/semantics/types/user.py]: Moved to a helper function

File: vyper/semantics/types/user.py [hunk 5]
```diff
@@ -415,31 +436,7 @@ def _ctor_call_return(self, node: vy_ast.Call) -> "StructT":
...
-                    kwarg,
-                )
-            expected_type = members.pop(argname)
-            validate_expected_type(kwarg.value, expected_type)
-
-        if members:
-            raise VariableDeclarationException(
-                f"Struct declaration does not define all fields: {', '.join(list(members))}", node
-            )
+        validate_kwargs(node, self.member_types.copy(), self.typeclass)
```
Comment (charles-cooper) [inline vyper/semantics/types/user.py:439]: Maybe better to copy inside validate_kwargs

Comment (z80dev) [reply vyper/semantics/types/user.py]: moved the copy into the fn, now the fn can't modify passed in member dicts which is definitely better, thanks

**Comment by z80dev** [comment]
merging in master dropped the coverage a good amount but from looking at the report I don't think its related to this PR, but will look into it

**Comment by charles-cooper** [comment]
> merging in master dropped the coverage a good amount but from looking at the report I don't think its related to this PR, but will look into it

i wouldn't worry about it too much, the codecov app often fails to merge the coverage files properly

File: vyper/semantics/analysis/utils.py [hunk 6]
```diff
@@ -694,3 +696,33 @@ def get_expr_writes(node: vy_ast.VyperNode) -> OrderedSet[VarAccess]:
         ret |= get_expr_writes(c)
     node._metadata["writes_r"] = ret
     return ret
+
+
+def validate_kwargs(node: vy_ast.Call, members: Dict, typeclass: str):
```
Comment (charles-cooper) [inline vyper/semantics/analysis/utils.py:701]:
```suggestion
def validate_kwargs(node: vy_ast.Call, members: dict[str, VyperType], typeclass: str):
```

**Review by charles-cooper — APPROVED** [review]
LGTM. Thank you!!

**Pull request merged**